feat(parser): improve diagnostic messages for merge conflicts

camc314 opened the pull request:
Adds detection and improved error messages for version control merge conflicts encountered in the parser.

This is based upon rustc's implementation.

camc314 commented on crates/oxc_parser/src/error_handler.rs:
```diff
@@ -71,3 +81,237 @@ impl<'a> ParserImpl<'a> {
+        if !token.is_on_new_line() && span.start != 0 {
+            return None;
+        }
+
```
yes probably, only case i can think of is if it was inside a multiline string/template string, but i think that it's enough of an edge case that it's fine

sapphi-red commented on crates/oxc_parser/src/diagnostics.rs:
```diff
@@ -38,6 +38,48 @@ pub fn unexpected_token(span: Span) -> OxcDiagnostic {
+    middle_span: Option<Span>,
+    end_span: Option<Span>,
+) -> OxcDiagnostic {
+    let mut diagnostic = OxcDiagnostic::error("encountered diff marker")
```
```suggestion
    let mut diagnostic = OxcDiagnostic::error("Encountered diff marker")
```
other errors have the title capitialized.

sapphi-red commented on crates/oxc_parser/src/diagnostics.rs:
```diff
@@ -38,6 +38,48 @@ pub fn unexpected_token(span: Span) -> OxcDiagnostic {
+         already had and the bottom section is the remote code\n\
+         if you're in the middle of a rebase, the top section is the code being rebased onto and \
+         the bottom section is the code coming from the current commit being rebased\n\
+         if you have nested conflicts, resolve the outermost conflict first",
```
```suggestion
            "Conflict markers indicate that a merge was started but could not be completed due to \
         merge conflicts.\n\
         To resolve a conflict, keep only the code you want and then delete the lines containing \
         conflict markers.\n\
         If you're having merge conflicts after pulling new code, the top section is the code you \
         already had and the bottom section is the remote code.\n\
         If you're in the middle of a rebase, the top section is the code being rebased onto and \
         the bottom section is the code coming from the current commit being rebased.\n\
         If you have nested conflicts, resolve the outermost conflict first.",
```

sapphi-red commented on crates/oxc_parser/src/diagnostics.rs:
```diff
@@ -38,6 +38,48 @@ pub fn unexpected_token(span: Span) -> OxcDiagnostic {
+        // Incomplete conflict - missing middle or end markers
+        diagnostic = diagnostic.with_help(
+            "this conflict marker appears to be incomplete (missing `=======` or `>>>>>>>`)\n\
+         check if the conflict markers were accidentally modified or partially deleted",
```
```suggestion
            "This conflict marker appears to be incomplete (missing `=======` or `>>>>>>>`).\n\
         Check if the conflict markers were accidentally modified or partially deleted.",
```

camc314 committed:
feat(parser): improve diagnostic messages for merge conflicts (#15443)

Adds detection and improved error messages for version control merge conflicts encountered in the parser.

This is based upon rustc's implementation.
